File: packages/ld-adapter/src/provider/launch-darkly-provider.ts
import type {
	ApplyTargetingInput,
	GetFlagStateInput,
	LaunchDarklyClientConfig,
} from '@ff-promo/contracts';
import {
	createLaunchDarklyClient,
	type LaunchDarklyRawClient,
} from '../client/ld-api-client.js';
import {
	createRateLimitedLdClient,
	type RateLimitedLdClient,
} from '../client/rate-limited-client.js';
import { getFlagState } from '../read/get-flag-state.js';
import type { FlagProvider } from './flag-provider.js';
import { applyTargeting } from '../write/apply-targeting.js';

export class LaunchDarklyProvider implements FlagProvider {
	constructor(
		private readonly rawClient: LaunchDarklyRawClient,
		private readonly rateLimitedClient: RateLimitedLdClient,
	) {}

	getFlagState(input: GetFlagStateInput) {
		return this.rateLimitedClient.schedule(() =>
			getFlagState({ client: this.rawClient }, input),
		);
	}

	applyTargeting(input: ApplyTargetingInput) {
		return applyTargeting({ rateLimitedClient: this.rateLimitedClient }, input);
	}
}

export function createLaunchDarklyProvider(
	config: LaunchDarklyClientConfig,
): LaunchDarklyProvider {
	const rawClient = createLaunchDarklyClient(config);
	const rateLimitedClient = createRateLimitedLdClient(rawClient);
	return new LaunchDarklyProvider(rawClient, rateLimitedClient);
}

File: kotlin/modules/ld-adapter/src/main/kotlin/com/ffpromo/ldadapter/provider/LaunchDarklyProvider.kt
package com.ffpromo.ldadapter.provider

import com.ffpromo.contracts.ApplyTargetingInput
import com.ffpromo.contracts.GetFlagStateInput
import com.ffpromo.contracts.LaunchDarklyClientConfig
import com.ffpromo.ldadapter.client.LaunchDarklyRawClient
import com.ffpromo.ldadapter.client.RateLimitedLdClient
import com.ffpromo.ldadapter.client.createLaunchDarklyClient
import com.ffpromo.ldadapter.client.createRateLimitedLdClient
import com.ffpromo.ldadapter.read.GetFlagStateDeps
import com.ffpromo.ldadapter.read.getFlagState
import com.ffpromo.ldadapter.write.ApplyTargetingDeps
import com.ffpromo.ldadapter.write.applyTargeting

class LaunchDarklyProvider(
    private val rawClient: LaunchDarklyRawClient,
    private val rateLimitedClient: RateLimitedLdClient,
) : FlagProvider {
    override suspend fun getFlagState(input: GetFlagStateInput) =
        rateLimitedClient.schedule {
            getFlagState(GetFlagStateDeps(rawClient), input)
        }

    override suspend fun applyTargeting(input: ApplyTargetingInput) =
        applyTargeting(ApplyTargetingDeps(rateLimitedClient), input)
}

fun createLaunchDarklyProvider(config: LaunchDarklyClientConfig): LaunchDarklyProvider {
    val rawClient = createLaunchDarklyClient(config)
    val rateLimitedClient = createRateLimitedLdClient(rawClient)
    return LaunchDarklyProvider(rawClient, rateLimitedClient)
}
